x/ccv/provider/types/msg.go

```diff
@@ -93,3 +109,99 @@ func ParseConsumerKeyFromJson(jsonStr string) (pkType, key string, err error) {
 	}
 	return pubKey.Type, pubKey.Key, nil
 }
+
+func NewMsgSubmitConsumerMisbehaviour(submitter sdk.AccAddress, misbehaviour *ibctmtypes.Misbehaviour) (*MsgSubmitConsumerMisbehaviour, error) {
+	return &MsgSubmitConsumerMisbehaviour{Submitter: submitter.String(), Misbehaviour: misbehaviour}, nil
+}
+
+// Route implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerMisbehaviour) Route() string { return RouterKey }
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerMisbehaviour) Type() string {
+	return TypeMsgSubmitConsumerMisbehaviour
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerMisbehaviour) ValidateBasic() error {
+	if msg.Submitter == "" {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidAddress, msg.Submitter)
+	}
+
+	if err := msg.Misbehaviour.ValidateBasic(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerMisbehaviour) GetSignBytes() []byte {
+	bz := ccvtypes.ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerMisbehaviour) GetSigners() []sdk.AccAddress {
+	addr, err := sdk.AccAddressFromBech32(msg.Submitter)
+	if err != nil {
+		// same behavior as in cosmos-sdk
+		panic(err)
+	}
+	return []sdk.AccAddress{addr}
+}
+
+func NewMsgSubmitConsumerDoubleVoting(submitter sdk.AccAddress, ev *tmtypes.DuplicateVoteEvidence, header *ibctmtypes.Header) (*MsgSubmitConsumerDoubleVoting, error) {
+	return &MsgSubmitConsumerDoubleVoting{Submitter: submitter.String(), DuplicateVoteEvidence: ev, InfractionBlockHeader: header}, nil
+}
+
+// Route implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) Route() string { return RouterKey }
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) Type() string {
+	return TypeMsgSubmitConsumerDoubleVoting
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) ValidateBasic() error {
```

Comment:
How thorough do we have to be with these checks?
Could a malicious actor that provides `nil` to a message field that we do not check against halt the chain?


E..g, Should we check that `VoteA` and `VoteB` are also there and that each have a `BlockID` that is not `nil` and so on and so forth?

Reply:
It can be argued to have a stricter validation. However, any panic will be recovered and will just reject the tx without halting the chain.

Reply:
Thanks. I wasn't aware of this. Resolving it.